src/tss2-tcti/tcti-helper-common.c

```diff
@@ -0,0 +1,761 @@
+
+    for (int retries = 100; retries > 0; retries--) {
+        /* In case of failed read div_vid is set to zero */
+        ctx->read_reg (ctx->data, TCTI_HELPER_COMMON_REG_TPM_DID_VID, &did_vid, sizeof (did_vid));
+        if (did_vid != 0) break;
+        /* TPM might be resetting, let's retry in a bit */
+        rc = ctx->sleep_ms (ctx->data, POLLING_INTERVAL_MS);
+        if (rc != TSS2_RC_SUCCESS) {
+            LOG_ERROR ("Failed to invoke sleep_ms()");
+            return rc;
+        }
+    }
+    if (did_vid == 0) {
+        LOG_ERROR ("Probing TPM failed");
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+    LOG_DEBUG ("Probing TPM successful");
+
+    if (is_i2c) {
+        /* Read i2c guard time */
+        LOG_DEBUG ("Reading the i2c GUARD_TIME value");
+        rc = tcti_helper_common_read_guard_time (ctx);
+        if (rc != TSS2_RC_SUCCESS) {
+            return TSS2_TCTI_RC_IO_ERROR;
+        }
+
+        /* Enable Data Checksum */
+        LOG_DEBUG ("Enable Data Checksum");
+        rc = tcti_helper_common_enable_crc (ctx);
+        if (rc != TSS2_RC_SUCCESS) {
+            return TSS2_TCTI_RC_IO_ERROR;
+        }
+    }
+
+    ctx->is_i2c = is_i2c;
+
+    /* Claim locality */
+    LOG_DEBUG ("Claiming TPM locality...");
+    rc = tcti_helper_common_claim_locality (ctx);
+    if (rc != TSS2_RC_SUCCESS) {
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    /* Wait up to TIMEOUT_B for TPM to become ready */
+    LOG_DEBUG ("Waiting for TPM to become ready...");
+    rc = tcti_helper_common_wait_for_status (ctx, expected_status_bits, expected_status_bits, TIMEOUT_B);
+    if (rc == TSS2_TCTI_RC_TRY_AGAIN) {
+        /*
+         * TPM did not auto transition into ready state,
+         * write 1 to commandReady to start the transition.
+         */
+        if ((rc = tcti_helper_common_write_sts_reg (ctx, TCTI_HELPER_COMMON_TPM_STS_COMMAND_READY))) {
+            LOG_ERROR ("Failed to write the commandReady bit to TPM_STS register");
+            return rc;
+        }
+        rc = tcti_helper_common_wait_for_status (ctx, expected_status_bits, expected_status_bits, TIMEOUT_B);
+    }
+
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to wait for the TPM to become ready");
+        return rc;
+    }
+
+    LOG_DEBUG ("TPM is ready");
+
+    /* Get rid */
+    rc = ctx->read_reg (ctx->data, TCTI_HELPER_COMMON_REG_TPM_RID, &rid, sizeof(rid));
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR ("Failed to read the TPM_RID register");
+        return rc;
+    }
+
+#if MAXLOGLEVEL >= LOGL_INFO
+    /* Print device details */
+    uint16_t vendor_id, device_id;
+    vendor_id = did_vid & 0xffff;
+    device_id = did_vid >> 16;
+    LOG_DEBUG ("Connected to TPM with vid:did:rid of 0x%04" PRIx16 ":%04" PRIx16 ":%02" PRIx8, vendor_id, device_id, rid);
+#endif
+
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC Tcti_Helper_Common_Transmit (TCTI_HELPER_COMMON_CONTEXT *ctx, size_t size, const uint8_t *cmd_buf)
```

Comment:
We are missing the retry-logic in this function.
It always goes directly to "return error"